File: backend/node/src/controllers/UploadImage.mjs
import { consult } from "../database/database.mjs";
import { UploadS3_, generateSignedUrl } from "../aws/S3Privado.mjs";
import fs from "fs";

const Upload = async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ error: "No se ha subido ningún archivo" });
    }
    const filePath = req.file.path; // Ruta temporal del archivo
    const fileName = req.file.filename; // Nombre temporal del archivo
    const fileType = req.file.mimetype; // Tipo MIME del archivo
    const OriginalName = req.file.originalname; // Nombre original del archivo
    //const size = req.file.size; // Tamaño del archivo en bytes
    const CarpetaS3 = "Album_seminario1";

    const { album_id, descripcion, usuario_id } = req.body;

    if (!album_id || !descripcion || !usuario_id) {
      fs.unlinkSync(filePath);
      return res.status(400).json({ message: "Faltan datos" });
    }

    //no se puede subir al album de fotos de perfil
    const fotos_perfil = await consult(
      `SELECT MIN(id) AS id FROM album WHERE usuario_id = ${usuario_id};`
    );
    if (fotos_perfil[0].status !== 200 || fotos_perfil[0].result.length === 0) {
      fs.unlinkSync(filePath);
      return res
        .status(404)
        .json({ message: "No se ha podido subir el archivo" });
    }

    if (fotos_perfil[0].result[0].id == album_id) {
      fs.unlinkSync(filePath);
      return res.status(400).json({
        message: "No se puede subir imagenes al album de fotos de perfil",
      });
    }

    //despues de subir a s3, elimino el archivo temporal
    const response = await UploadS3_(filePath, fileName, fileType, CarpetaS3);
    if (!response) {
      return res.status(500).json({ error: "Error al subir el archivo a S3" });
    }

    const fecha = new Date().toISOString().slice(0, 19).replace("T", " ");
    const url = `${CarpetaS3}/${fileName}`;
    const query =
      await consult(`INSERT INTO imagen (album_id, nombre, url_s3, descripcion, fecha_creacion) 
    VALUES (${album_id}, '${OriginalName}', '${url}', '${descripcion}', '${fecha}');`);

    if (query[0].status !== 200) {
      console.error(
        "Error al subir el archivo a la base de datos:",
        query[0].message
      );
      return res
        .status(500)
        .json({ message: "Error al subir el archivo a la base de datos" });
    }

    res.status(200).json({ message: "Archivo subido correctamente" });
  } catch (error) {
    console.error("Error al eliminar el album:", error);
    res.status(500).json({ message: "Error al eliminar el album" });
  }
};

const Get = async (req, res) => {
  try {
    const { album_id } = req.body;

    if (!album_id) {
      return res.status(400).json({ message: "Faltan datos" });
    }

    const response = await consult(`SELECT * FROM imagen WHERE album_id = ${album_id};`);

    if (response[0].status !== 200) {
      return res
        .status(404)
        .json({ message: "No se encontraron archivos en el album" });
    }

    const images = response[0].result;
    for (const image of images) {
      if (image.url_s3 !== null) {
        image.url_s3 = await generateSignedUrl(image.url_s3);
      }
    }
    res.status(200).json(images);
  } catch (error) {
    console.error("Error al obtener los archivos del album:", error);
    res
      .status(500)
      .json({ message: "Error al obtener los archivos del album" });
  }
};

const Gallery = async (req, res) => {
  try {
    const { usuario_id } = req.body;

    if (!usuario_id) {
      return res.status(400).json({ message: "Faltan datos" });
    }

    const query = `
    SELECT 
      a.id AS album_id,
      a.nombre AS album_nombre,
      (select url_s3 from imagen where album_id= i.album_id Limit 1) as url,
      COUNT(i.id) AS cantidad_imagenes
    FROM 
      album a
    LEFT JOIN 
      imagen i ON a.id = i.album_id
    WHERE 
      a.usuario_id = ${usuario_id}
    GROUP BY 
      a.id;`;

    const response = await consult(query);

    if (response[0].status !== 200) {
      return res
        .status(404)
        .json({ message: "No se encontraron archivos en el album" });
    }

    let albums = response[0].result;
    for (const album of albums) {
      if (album.url !== null) {
        album.url = await generateSignedUrl(album.url); // Ahora sí se espera el await
      }
    }
    res.status(200).json(albums);
  } catch (error) {
    console.error("Error al obtener los archivos del album:", error);
    res
      .status(500)
      .json({ message: "Error al obtener los archivos del album" });
  }
};

export const UploadImage = {
  Upload,
  Get,
  Gallery,
};

File: backend/python/src/controllers/UploadImage.py
from database.database import consult
from aws.S3Privado import generate_signed_url, upload_s3
from flask import request, jsonify
import os
from werkzeug.utils import secure_filename
import datetime

UPLOAD_FOLDER = 'uploads/'

def Upload():
    try:
        if 'file' not in request.files:
            return jsonify({"message": "No se ha enviado ninguna imagen"}), 400
        
        data = request.form
        album_id = data['album_id']
        descripcion = data['descripcion']
        usuario_id = data['usuario_id']

        if not album_id or not descripcion or not usuario_id:
            return jsonify({"message": "Faltan datos"}), 400
        
        if not os.path.exists(UPLOAD_FOLDER):
                os.makedirs(UPLOAD_FOLDER)

        file = request.files['file']
        filename = secure_filename(file.filename)
        file_path = os.path.join(UPLOAD_FOLDER, filename)
        file.save(file_path)

        filetype = file.mimetype

        CarpetaS3 = "Fotos_Publicadas"

        fotos_perfil  = consult(f"SELECT MIN(id) as id FROM album WHERE usuario_id = {usuario_id}")

        if fotos_perfil[0]['status'] != 200 or len(fotos_perfil[0]['result']) == 0:
            os.remove(file_path)
            return jsonify({"message": "No se ha podido subir la imagen"}), 404
        
        if int(fotos_perfil[0]['result'][0]['id']) == int(album_id):
            os.remove(file_path)
            return jsonify({"message": "No se puede subir la imagen al album de fotos de perfil"}), 400
        
        response = upload_s3(file_path, filename, filetype, CarpetaS3)
        if not response:
            os.remove(file_path)
            return jsonify({"message": "Error al subir imagen"}), 500
        
        url = f"{CarpetaS3}/{filename}"
        query = consult(f"INSERT INTO imagen (album_id, nombre, url_s3, descripcion, fecha_creacion) VALUES ({album_id}, '{filename}', '{url}', '{descripcion}', CURRENT_TIMESTAMP);")
        
        if query[0]['status'] != 200:
            return jsonify({"message": "Error al subir imagena la base de datos"}), 500
        
        return jsonify({"message":"Imagen subida correctamente"}), 200
    except Exception as e:
        print("Error al subir imagen" + e)
        return jsonify({"message": "Error al subir imagen"}), 500
    
def Get():
    try:
        data = request.json
        album_id = data['album_id']

        if not album_id:
            return jsonify({"message": "Faltan datos"}), 400
        
        query = consult(f"SELECT * FROM imagen WHERE album_id = {album_id};")

        if query[0]['status'] != 200:
            return jsonify({"message": "No se encontraron archivos en el album"}), 400
        
        images = query[0]['result']

        for image in images:
            image['url_s3'] = generate_signed_url(image['url_s3'])

        return jsonify(images), 200
                
    except Exception as e:
        print("Error al obtener las imagenes" + e)
        return jsonify({"message": "Error al obtener las imagenes"}), 500
    
def Gallery():
    try:
        data = request.json
        usuario_id = data['usuario_id']

        if not usuario_id:
            return jsonify({"message": "Faltan datos"}), 400
        
        query = consult(f"SELECT a.id AS album_id, a.nombre AS album_nombre, (select url_s3 from imagen where album_id = i.album_id limit 1) as url, COUNT(i.id) as cantidad_imagenes FROM album a LEFT JOIN imagen i ON a.id = i.album_id WHERE a.usuario_id = {usuario_id} GROUP BY a.id;")

        if query[0]['status'] != 200:
            return jsonify({"message": "No se encontraron archivos en el album"}), 500
        
        albums = query[0]['result']

        for album in albums:
            if album['url']:
                album['url'] = generate_signed_url(album['url'])
        
        return jsonify(albums), 200
    except Exception as e:
        print("Error al obtener los archivos del album" + e)
        return jsonify({"message": "Error al obtener los album"}), 500
    
UploadImage = {
    "Upload": Upload,
    "Get": Get,
    "Gallery": Gallery
}